src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,140 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static int COUNT = 2;
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+
     @Override
     public void add(T value) {
+        addLast(value);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+
+        if (index == size) {
+            addLast(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(value, current.getPrev(), current);
+            if (current.getPrev() != null) {
+                current.getPrev().setNext(newNode);
+            } else {
+                first = newNode;
+            }
+            current.setPrev(newNode);
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            addLast(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+        }
+        return getNode(index).getData();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = getNode(index);
+        T oldValue = current.getData();
+        current.setData(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = getNode(index);
+        if (current.getPrev() == null) {
+            first = current.getNext();
+        } else {
+            current.getPrev().setNext(current.getNext());
+        }
+        if (current.getNext() != null) {
+            current.getNext().setPrev(current.getPrev());
+        } else {
+            last = current.getPrev();
+        }
```

Comment:
duplicated logic in two remove methods, extract it to unlink method
![image](https://github.com/mate-academy/jv-linked-list/assets/49767049/99d6a279-d420-41d1-85a5-8507152ffb1c)